tests/core/test_block_manager.py

```diff
@@ -85,7 +85,43 @@ def test_allocate():
                                         watermark=1 / num_gpu_blocks)
     for i in range(num_gpu_blocks - 1):
         _, seq_group = create_dummy_prompt(str(i), block_size)
-        assert block_manager.can_allocate(seq_group)
+        assert block_manager.can_allocate(seq_group) == AllocStatus.OK
+        block_manager.allocate(seq_group)
+    assert block_manager.can_allocate(seq_group) != AllocStatus.OK
+
+
+def test_allocate_encoder_decoder():
+    block_size = 4
+    num_cpu_blocks = 4
+    num_gpu_blocks = 4
+    block_req_per_seq_group = 2
+    block_manager = BlockSpaceManagerV1(block_size,
+                                        num_cpu_blocks,
+                                        num_gpu_blocks,
+                                        watermark=0)
+
+    # Allocate same sequence group to all available gpu blocks.
+    for i in range(num_gpu_blocks // block_req_per_seq_group):
+        _, _, seq_group = create_dummy_prompt_encoder_decoder(
+            str(i),
+            decoder_prompt_length=block_size,
+            encoder_prompt_length=block_size)
+        assert block_manager.can_allocate(seq_group) == AllocStatus.OK
+        block_manager.allocate(seq_group)
+    assert block_manager.can_allocate(seq_group) != AllocStatus.OK
+
+    # Allocate same sequence group to all available gpu blocks.
+    # Use watermark to reserve one gpu block.
+    block_manager = BlockSpaceManagerV1(block_size,
```

Comment:
wondering if it makes sense to extend the test with one more case  with enable_caching as true and verify that the allocation fails?

Reply:
Done:
* Block manager v1 now has two new tests, which respectively confirm that NotImplemented errors are raised when prefix caching or sliding window attention (SWA) are enabled when an encoder/decoder SequenceGroup is passed to the BlockManager via allocate() or can_allocate()
* Block manager v2 now has two new tests
    * One test confirms that a NotImplemented error is raised when prefix caching is enabled in the block manager constructor, followed by an encoder/decoder SequenceGroup being passed to the BlockManager via can_allocate()
     * As for SWA, block manager v2 does not yet support SWA & already fails with an assertion in the constructor if a SWA window size is passed in. However, even when decoder-only SWA is eventually implemented, it is likely that encoder/decoder SWA may be a separate workstream (or not prioritized at all.) Therefore, I added a test which only passes if either (1) the block manager v2 constructor fails with AssertionError when a SWA window size is passed (current state of affairs), or (2) can_allocate() fails with NotImplementedError when an encoder/decoder SequenceGroup is passed in.